cyral/internal/policyv2/model.go

```diff
@@ -0,0 +1,163 @@
+	Enforced    bool       `json:"enforced,omitempty"`
+	Type        string     `json:"type,omitempty"`
+}
+
+// WriteToSchema writes the policy data to the schema
+func (r PolicyV2) WriteToSchema(d *schema.ResourceData) error {
+	if err := d.Set("id", r.Policy.ID); err != nil {
+		return fmt.Errorf("error setting 'id' field: %w", err)
+	}
+	if err := d.Set("name", r.Policy.Name); err != nil {
+		return fmt.Errorf("error setting 'name' field: %w", err)
+	}
+	if err := d.Set("description", r.Policy.Description); err != nil {
+		return fmt.Errorf("error setting 'description' field: %w", err)
+	}
+	if err := d.Set("enabled", r.Policy.Enabled); err != nil {
+		return fmt.Errorf("error setting 'enabled' field: %w", err)
+	}
+	if err := d.Set("tags", r.Policy.Tags); err != nil {
+		return fmt.Errorf("error setting 'tags' field: %w", err)
+	}
+	if err := d.Set("valid_from", r.Policy.ValidFrom); err != nil {
+		return fmt.Errorf("error setting 'valid_from' field: %w", err)
+	}
+	if err := d.Set("valid_until", r.Policy.ValidUntil); err != nil {
+		return fmt.Errorf("error setting 'valid_until' field: %w", err)
+	}
+
+	// Normalize the document field before setting it
+	normalizedDocument, err := normalizeJSON(r.Policy.Document)
+	if err != nil {
+		return fmt.Errorf("error normalizing 'document' field: %w", err)
+	}
+	if err := d.Set("document", normalizedDocument); err != nil {
+		return fmt.Errorf("error setting 'document' field: %w", err)
+	}
+
+	if err := d.Set("last_updated", map[string]interface{}{
+		"actor":      r.Policy.LastUpdated.Actor,
+		"actor_type": r.Policy.LastUpdated.ActorType,
+		"timestamp":  r.Policy.LastUpdated.Timestamp,
+	}); err != nil {
+		return fmt.Errorf("error setting 'last_updated' field: %w", err)
+	}
+	if err := d.Set("created", map[string]interface{}{
+		"actor":      r.Policy.Created.Actor,
+		"actor_type": r.Policy.Created.ActorType,
+		"timestamp":  r.Policy.Created.Timestamp,
+	}); err != nil {
+		return fmt.Errorf("error setting 'created' field: %w", err)
+	}
+	if err := d.Set("enforced", r.Policy.Enforced); err != nil {
+		return fmt.Errorf("error setting 'enforced' field: %w", err)
+	}
+	if r.Policy.Type != "" {
+		if err := d.Set("type", r.Policy.Type); err != nil {
+			return fmt.Errorf("error setting 'type' field: %w", err)
+		}
+	}
+	d.SetId(r.Policy.ID)
+	return nil
+}
+
+// ReadFromSchema reads the policy data from the schema
+func (r *PolicyV2) ReadFromSchema(d *schema.ResourceData) error {
+	r.Policy.ID = d.Get("id").(string)
+	r.Policy.Name = d.Get("name").(string)
+	r.Policy.Description = d.Get("description").(string)
+	r.Policy.Enabled = d.Get("enabled").(bool)
+	r.Policy.Tags = expandStringList(d.Get("tags").([]interface{}))
+	r.Policy.ValidFrom = d.Get("valid_from").(string)
+	r.Policy.ValidUntil = d.Get("valid_until").(string)
+
+	// Normalize the document field before storing it
+	document := d.Get("document").(string)
+	normalizedDocument, err := normalizeJSON(document)
```

Comment:
As soon as we do this, we lose the original value of document provided by the resource definition (whitespace changes) - this may be the cause of the state mismatch that you're seeing.

Reply:
code was refactored